docs/proposals/super-simple-discovery-api.md

```diff
@@ -0,0 +1,205 @@
+* `kubeadm` verifies the JWS and can now trust the server.  Further
+  communication is simpler as the CA certificate in the kubeconfig file can be
+  trusted.
+
+
+#### NEW: Token Structure
+
+To first make this work, we put some structure into the token.  It has both a token identifier and the token value, separated by a dot.  Example:
+
+```
+ae23dc.faddc87f5a5ab458
+```
+
+The first part of the token is the `token-id`.  The second part is the `token-secret`.  By having a token identifier, we make it easier to specify *which* token you are talking about without sending the token itself in the clear.
+
+To support this new type of token, we will update the CSV token authenticator to recognize these types of tokens and index them by `token-id`.  Tokens that do not conform to this method will continue to work but won't be usable using the JWT based authentication mechanism defined below.
+
+#### Quick Primer on JWS and JWT
+
+[JSON Web Signatures](https://tools.ietf.org/html/rfc7515) are a way to sign, serialize and verify a payload.  It supports both symmetric keys (aka shared secrets) along with asymmetric keys (aka public key infrastructure or key pairs).  The JWS is split in to 3 parts:
+1. a header about how it is signed
+2. the clear text payload
+3. the signature.
+
+There are a couple of different ways of encoding this data -- either as a JSON object or as a set of BASE64URL strings for including in headers or URL parameters.  In this case, we are using a shared secret and the HMAC-SHA256 signing algorithm and encoding it as a JSON object.  The popular JWT (JSON Web Tokens) specification is a type of JWS.
+
+The JWS specification [describes how to encode](https://tools.ietf.org/html/rfc7515#appendix-F) "detached content".  In this way the signature is calculated as normal but the content isn't included in the signature.
+
+#### NEW: JWT authentication
+
+To avoid passing an unencoded token we will instead create a short lived JWT based on the token.  The JWT will be constructed as so:
+
+* **Header**
+  * **typ**: `JWT`. REQUIRED.
+  * **alg**: `HS256` for HMAC SHA-256. REQUIRED.
+  * **kid**: The token-id. REQUIRED.
+* **Claims**
+  * **sub**: The token-id. REQUIRED.
+  * **aud**: The long term globally unique cluster identifier.  This is not implemented yet so is OPTIONAL.
+  * **exp**: An expiration time for this JWT.  This must be no more than 2 minutes in the future. REQUIRED.
+  * **iat**: The time that the JWT was issued.  OPTIONAL.
+
+**Question:** The 2 minute value is picked to be short enough to discourage replay attacks without introducing the complexity of tracking nonces.  If we mandated the use of nonces, we'd have to track which nonces were used across HA instances.  Is there a better value than 2 minutes?  If we see clock skew greater or connectivity problems greater than 2 minutes we'll probably have other problems.
```

Comment:
Since we never need to use this token except on secured channels, I don't think we need to make the token a full JWT - you can just use the secret value as a "master token".  I don't think the nonce adds anything unless you don't trust TLS, and if you don't trust TLS a 2 minute timeout isn't going to help you.

That avoids needing anything other than a rotatable cluster key.